File: CS/Form1.cs
using System;
using System.Data;
using System.Windows.Forms;
using DevExpress.XtraScheduler;
using System.Collections.Generic;

namespace SchedulerFindAptByRow {
    public partial class Form1 : Form {
        Dictionary<object, Appointment> appointmentRowDictionary = new Dictionary<object, Appointment>();
        Dictionary<Appointment, object> rowAppointmentDictionary = new Dictionary<Appointment, object>();

        public Form1() {
            InitializeComponent();
        }

        protected Dictionary<object, Appointment> AppointmentRowDictionary { get { return appointmentRowDictionary; } }
        protected Dictionary<Appointment, object> RowAppointmentDictionary { get { return rowAppointmentDictionary; } }

        private void Form1_Load(object sender, EventArgs e) {
            // TODO: This line of code loads data into the 'carsDBDataSet.Cars' table. You can move, or remove it, as needed.
            this.carsTableAdapter.Fill(this.carsDBDataSet.Cars);
            // TODO: This line of code loads data into the 'carsDBDataSet.CarScheduling' table. You can move, or remove it, as needed.
            this.carSchedulingTableAdapter.Fill(this.carsDBDataSet.CarScheduling);

            schedulerStorage1.Resources.DataSource = carsDBDataSet.Cars;
            schedulerStorage1.Appointments.DataSource = carsDBDataSet.CarScheduling;

            dataGridView1.DataSource = carsDBDataSet.CarScheduling;

            if (schedulerStorage1.Appointments.Count > 0)
                schedulerControl1.Start = schedulerStorage1.Appointments[0].Start;
        }

        private void schedulerStorage1_AppointmentCollectionLoaded(object sender, EventArgs e) {
            foreach (Appointment apt in schedulerStorage1.Appointments.Items) {
                AppointmentBaseCollection apts = new AppointmentBaseCollection();
                apts.Add(apt);
                RegisterAppointmentRows(apts);
                if (apt.Type == AppointmentType.Pattern) {
                    RegisterAppointmentRows(apt.GetExceptions());
                }
            }
        }

        private void schedulerStorage1_AppointmentCollectionCleared(object sender, EventArgs e) {
            AppointmentRowDictionary.Clear();
            RowAppointmentDictionary.Clear();
        }

        private void schedulerStorage1_AppointmentsInserted(object sender, PersistentObjectsEventArgs e) {
            AppointmentBaseCollection apts = (AppointmentBaseCollection)e.Objects;
            foreach (Appointment apt in apts) {
                DataRow row = GetAppointmentRow(apt);
                RegisterAppointmentRow(row, apt);
            }
        }

        private void schedulerStorage1_AppointmentsDeleted(object sender, PersistentObjectsEventArgs e) {
            AppointmentBaseCollection apts = (AppointmentBaseCollection)e.Objects;
            foreach (Appointment apt in apts) {
                DataRow row = (DataRow)RowAppointmentDictionary[apt];
                AppointmentRowDictionary.Remove(row);
                RowAppointmentDictionary.Remove(apt);
            }
        }

        private void RegisterAppointmentRows(AppointmentBaseCollection apts) {
            foreach (Appointment apt in apts) {
                DataRow row = GetAppointmentRow(apt);
                RegisterAppointmentRow(row, apt);
            }
        }

        private void RegisterAppointmentRow(DataRow row, Appointment apt) {
            if (!AppointmentRowDictionary.ContainsKey(row)) {
                AppointmentRowDictionary.Add(row, apt);
            }
            if (!RowAppointmentDictionary.ContainsKey(apt)) {
                RowAppointmentDictionary.Add(apt, row);
            }
        }

        private DataRow GetAppointmentRow(Appointment apt) {
            return ((DataRowView)schedulerStorage1.GetObjectRow(apt)).Row;
        }

        protected Appointment FindAppointmentByRow(DataRow row) {
            if (AppointmentRowDictionary.ContainsKey(row))
                return AppointmentRowDictionary[row];
            else
                return null;
        }

        private void dataGridView1_SelectionChanged(object sender, EventArgs e) {
            UpdateLabelVisible(false);
            if (dataGridView1.SelectedRows.Count != 0) {
                int rowIndex = dataGridView1.Rows.IndexOf(dataGridView1.SelectedRows[0]);
                DataRow row = this.carsDBDataSet.CarScheduling.Rows[rowIndex];
                Appointment apt = FindAppointmentByRow(row);

                schedulerControl1.Services.AppointmentSelection.ClearSelection();
                if (apt != null) {
                    if (apt.Type == AppointmentType.Pattern || apt.Type == AppointmentType.DeletedOccurrence) {
                        UpdateLabelVisible(true);
                        return;
                    }
                    schedulerControl1.ActiveView.SelectAppointment(apt);
                }
            }
        }

        private void schedulerControl1_Click(object sender, EventArgs e) {
            UpdateLabelVisible(false);
        }

        private void UpdateLabelVisible(bool visible) {
            lbNoAppointment.Visible = visible;
        }

        private void dataGridView1_CellFormatting(object sender, DataGridViewCellFormattingEventArgs e) {
            if (dataGridView1.Columns[e.ColumnIndex].DataPropertyName == "EventType")
                e.Value = ((AppointmentType)Convert.ToInt32(e.Value)).ToString();
        }

        private void dataGridView1_CellValueNeeded(object sender, DataGridViewCellValueEventArgs e) {
            if (dataGridView1.Columns[e.ColumnIndex].Name == "Number")
                e.Value = e.RowIndex;
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler
Imports System.Collections.Generic

Namespace SchedulerFindAptByRow

    Public Partial Class Form1
        Inherits Form

        Private appointmentRowDictionaryField As Dictionary(Of Object, Appointment) = New Dictionary(Of Object, Appointment)()

        Private rowAppointmentDictionaryField As Dictionary(Of Appointment, Object) = New Dictionary(Of Appointment, Object)()

        Public Sub New()
            InitializeComponent()
        End Sub

        Protected ReadOnly Property AppointmentRowDictionary As Dictionary(Of Object, Appointment)
            Get
                Return appointmentRowDictionaryField
            End Get
        End Property

        Protected ReadOnly Property RowAppointmentDictionary As Dictionary(Of Appointment, Object)
            Get
                Return rowAppointmentDictionaryField
            End Get
        End Property

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            ' TODO: This line of code loads data into the 'carsDBDataSet.Cars' table. You can move, or remove it, as needed.
            carsTableAdapter.Fill(carsDBDataSet.Cars)
            ' TODO: This line of code loads data into the 'carsDBDataSet.CarScheduling' table. You can move, or remove it, as needed.
            carSchedulingTableAdapter.Fill(carsDBDataSet.CarScheduling)
            schedulerStorage1.Resources.DataSource = carsDBDataSet.Cars
            schedulerStorage1.Appointments.DataSource = carsDBDataSet.CarScheduling
            dataGridView1.DataSource = carsDBDataSet.CarScheduling
            If schedulerStorage1.Appointments.Count > 0 Then schedulerControl1.Start = schedulerStorage1.Appointments(0).Start
        End Sub

        Private Sub schedulerStorage1_AppointmentCollectionLoaded(ByVal sender As Object, ByVal e As EventArgs)
            For Each apt As Appointment In schedulerStorage1.Appointments.Items
                Dim apts As AppointmentBaseCollection = New AppointmentBaseCollection()
                apts.Add(apt)
                RegisterAppointmentRows(apts)
                If apt.Type = AppointmentType.Pattern Then
                    RegisterAppointmentRows(apt.GetExceptions())
                End If
            Next
        End Sub

        Private Sub schedulerStorage1_AppointmentCollectionCleared(ByVal sender As Object, ByVal e As EventArgs)
            AppointmentRowDictionary.Clear()
            RowAppointmentDictionary.Clear()
        End Sub

        Private Sub schedulerStorage1_AppointmentsInserted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            Dim apts As AppointmentBaseCollection = CType(e.Objects, AppointmentBaseCollection)
            For Each apt As Appointment In apts
                Dim row As DataRow = GetAppointmentRow(apt)
                RegisterAppointmentRow(row, apt)
            Next
        End Sub

        Private Sub schedulerStorage1_AppointmentsDeleted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
            Dim apts As AppointmentBaseCollection = CType(e.Objects, AppointmentBaseCollection)
            For Each apt As Appointment In apts
                Dim row As DataRow = CType(RowAppointmentDictionary(apt), DataRow)
                AppointmentRowDictionary.Remove(row)
                RowAppointmentDictionary.Remove(apt)
            Next
        End Sub

        Private Sub RegisterAppointmentRows(ByVal apts As AppointmentBaseCollection)
            For Each apt As Appointment In apts
                Dim row As DataRow = GetAppointmentRow(apt)
                RegisterAppointmentRow(row, apt)
            Next
        End Sub

        Private Sub RegisterAppointmentRow(ByVal row As DataRow, ByVal apt As Appointment)
            If Not AppointmentRowDictionary.ContainsKey(row) Then
                AppointmentRowDictionary.Add(row, apt)
            End If

            If Not RowAppointmentDictionary.ContainsKey(apt) Then
                RowAppointmentDictionary.Add(apt, row)
            End If
        End Sub

        Private Function GetAppointmentRow(ByVal apt As Appointment) As DataRow
            Return CType(schedulerStorage1.GetObjectRow(apt), DataRowView).Row
        End Function

        Protected Function FindAppointmentByRow(ByVal row As DataRow) As Appointment
            If AppointmentRowDictionary.ContainsKey(row) Then
                Return AppointmentRowDictionary(row)
            Else
                Return Nothing
            End If
        End Function

        Private Sub dataGridView1_SelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateLabelVisible(False)
            If dataGridView1.SelectedRows.Count <> 0 Then
                Dim rowIndex As Integer = dataGridView1.Rows.IndexOf(dataGridView1.SelectedRows(0))
                Dim row As DataRow = carsDBDataSet.CarScheduling.Rows(rowIndex)
                Dim apt As Appointment = FindAppointmentByRow(row)
                schedulerControl1.Services.AppointmentSelection.ClearSelection()
                If apt IsNot Nothing Then
                    If apt.Type = AppointmentType.Pattern OrElse apt.Type = AppointmentType.DeletedOccurrence Then
                        UpdateLabelVisible(True)
                        Return
                    End If

                    schedulerControl1.ActiveView.SelectAppointment(apt)
                End If
            End If
        End Sub

        Private Sub schedulerControl1_Click(ByVal sender As Object, ByVal e As EventArgs)
            UpdateLabelVisible(False)
        End Sub

        Private Sub UpdateLabelVisible(ByVal visible As Boolean)
            lbNoAppointment.Visible = visible
        End Sub

        Private Sub dataGridView1_CellFormatting(ByVal sender As Object, ByVal e As DataGridViewCellFormattingEventArgs)
            If Equals(dataGridView1.Columns(e.ColumnIndex).DataPropertyName, "EventType") Then e.Value = CType(Convert.ToInt32(e.Value), AppointmentType).ToString()
        End Sub

        Private Sub dataGridView1_CellValueNeeded(ByVal sender As Object, ByVal e As DataGridViewCellValueEventArgs)
            If Equals(dataGridView1.Columns(e.ColumnIndex).Name, "Number") Then e.Value = e.RowIndex
        End Sub
    End Class
End Namespace
